toolkit/tools/scheduler/schedulerutils/preparerequest.go

```diff
@@ -182,36 +198,65 @@ func testNodesToRequests(pkgGraph *pkggraph.PkgGraph, buildState *GraphBuildStat
-// - It will check if the node corresponds to an entry in packagesToRebuild.
-// - It will check if all dependencies of the node were also cached. Exceptions:
-//   - "TypePreBuilt" nodes must use the cache and have no dependencies to check.
-func canUseCacheForNode(pkgGraph *pkggraph.PkgGraph, node *pkggraph.PkgNode, packagesToRebuild []*pkgjson.PackageVer, buildState *GraphBuildState) (canUseCache bool) {
-	// The "TypePreBuilt" nodes always use the cache.
-	if node.Type == pkggraph.TypePreBuilt {
-		canUseCache = true
-		return
-	}
-
-	// Check if the node corresponds to an entry in packagesToRebuild
+// isRequiredRebuild checks if a node is required to be rebuilt based  on the packagesToRebuild list.
+func isRequiredRebuild(node *pkggraph.PkgNode, packagesToRebuild []*pkgjson.PackageVer) (requiredRebuild bool) {
 	packageVer := node.VersionedPkg
-	canUseCache = !sliceutils.Contains(packagesToRebuild, packageVer, sliceutils.PackageVerMatch)
-	if !canUseCache {
-		logger.Log.Debugf("Marking node (type: %s, version: %s) for rebuild per user request", node.Type, packageVer)
-		return
+	requiredRebuild = sliceutils.Contains(packagesToRebuild, packageVer, sliceutils.PackageVerMatch)
+	if requiredRebuild {
+		logger.Log.Debugf("Marking (%s) for rebuild per user request", packageVer)
 	}
+	return
+}
 
-	// If any of the node's dependencies were built instead of being cached then a build is required.
+// canUseCacheForNode checks if the cache can be used for a given node by:
+//   - Assume the node is stale to begin (freshness == 0).
+//   - Check if all dependencies of the node were cached, and calculate the expected freshness of the node based on the freshest dependency.
+//   - If all dependencies are cached (freshness == 0, aka stale) then the node will keep freshness 0 and may use the cache.
+//   - If any dependency is fresh (aka freshness > 0) then the node can't use the cache and will inherit the freshness of
+//     the freshest dependency (possibly adjusted by -1 for certain edges).
+func canUseCacheForNode(pkgGraph *pkggraph.PkgGraph, node *pkggraph.PkgNode, buildState *GraphBuildState) (canUseCache bool, freshness int) {
+	freshness = 0
+	canUseCache = true
+
+	// If any of the node's dependencies were built instead of being cached then a build is required. We treat any node
+	// with a freshness > 0 as being built. Each layer of the build completed will decrement the freshness of the node by 1.
 	dependencies := pkgGraph.From(node.ID())
 	for dependencies.Next() {
 		dependency := dependencies.Node().(*pkggraph.PkgNode)
 
-		if !buildState.IsNodeCached(dependency) {
+		inheritedFreshness, shouldRebuild := calculateExpectedFreshness(dependency, buildState)
+		if inheritedFreshness > freshness {
+			freshness = inheritedFreshness
+		}
+
+		if shouldRebuild {
 			logger.Log.Debugf("Can't use cached version of %v because %v is rebuilding", node.FriendlyName(), dependency.FriendlyName())
```

Comment:
To consider: I'm thinking we might want to slightly update this message now. For instance, replace `is rebuilding` with `has been rebuilt and had a freshness of %d`.